### ISSUE TITLE
Accessibility Docs

### ISSUE THREAD

**Issue by TheSharpieOne** [reactstrap/reactstrap#1013]
We need accessibility docs which help explain how the end user of this library (the develop) can "close the gap" between what this library can and does provide in terms of accessibility and what is left up to the developer to implement in their specific usage.
For example, with things like `Tooltip` and `Popover` the "target" is outside of those components and thus cannot be affected by them, the developer will need to add `aria-describedby`/`aria-labeledby` and `aria-haspopup="true"` to the target element themselves. Similar, connecting `Label` and `Input` via the `for` prop on the `Label` and the `id` prop on the `Input`.
Bootstrap has something similar, but a little more generic: http://getbootstrap.com/docs/4.1/getting-started/accessibility/

**Comment by RockinRonE** [comment]
@TheSharpieOne Were you thinking separate accessibility docs like Bootstrap or something more integrated into each Component example? E.g. tooltips I can add line addressing the accessibility need for `aria-labelledby`

**Comment by TheSharpieOne** [comment]
Probably a little of each. A separate page like bootstrap's which explains that it is up to the developer to ensure accessibility. Also footnotes or something of specific pages explicitly calling out what the developer would need to do to make their usages of the component more accessible. This footnote wouldn't need to be anything more than a sentence and maybe code block:

## Accessibility 

Ensure you add `aria-describedby` or `aria-labeledby` to your target element with the value of the id on the `Tooltip`/`UncontrolledTooltip` like so:
```
<span aria-describedby="my-tooltip">Hover over me</span>
<UncontrolledTooltip id="my-tooltip">This is the tooltip value</span>
```

**Comment by deveshjadon98** [comment]
Hi, Is there any more help needed on accessibility docs? I would like to contribute on this.

**Comment by deveshjadon98** [comment]
@TheSharpieOne How do I give the messages to the reactstrap components for screen readers?

**Comment by TheSharpieOne** [comment]
Not sure what you mean. The few components which output complex DOM structure and contain icons use `aria-label` to provide content for screen readers (such as [`ModalHeader`](https://github.com/reactstrap/reactstrap/blob/1f1c0a89b022dfaaeaba970f113b3990374f478c/src/ModalHeader.js#L41-L43)).
There is also the `.sr-only` bootstrap class which will visually hide the content but still let it be read by a screen reader.

If you are talking about updating the reactstrap docs for components to add screen reader information, that is just in each of the component pages: https://github.com/reactstrap/reactstrap/tree/master/docs/lib/Components as well as a new page dedicated to an overview about accessibility.

**Comment by tamb** [comment]
@TheSharpieOne My proposal is publishing a sub dir `reactstrap/a11y` that the dev can import and it has a11y built in.  That way they can opt-in for fully accessibly components.  I'll being work on a simple one like 'Collapse' and get that going.  We can legit copy the a11y logic from jQuery bootstrap because that team does a fantastic job.

Personally, I think writing docs on how to make components accessible is just as much work as making them accessible themselves, so why not?

**Comment by TheSharpieOne** [comment]
The problem is that doing thing the right way in react makes it impossible to provide an a a11y component in a library type of system (without having the dev-user need to be the glue between separate component or making the components so rigid that they are less useful).

For instance, the example I posted above (and below) has 2 different components which are related to one-another. a11y states that the span should reference the tooltip's ID (and thus the tooltip needs an ID). If the dev-user doesn't provide an ID to the tooltip, we could generate one; sure. But that `span` would need to know about it and add the `aria-describedby`. You cannot have the tooltip component modify the DOM of the span to add that attribute either, that is anti-react and will easily break.
An `a11y` component may just take care of rendering both the `span` and the tooltip, but now they are tied together and it makes it more difficult for the dev-user to control the output. What if they want to use a `<p>` instead of `<span>` and what if they also want to control the element output for the tooltip? What about classes or any extra/additional props on each?
The component ends up having to deal with a lot of extra stuff because the dev-user couldn't give an `id` and `aria-describedby` prop.

```jsx
<span aria-describedby="my-tooltip">Hover over me</span>
<UncontrolledTooltip id="my-tooltip">This is the tooltip value</span>
```

All that being said, if you believe you have a good way to create a11y component which are still flexible and easy to consume, please make a PR. (I can kinda see how react's context can be used to solve some of the problems, but it may also create its own issues and/or make the components more cumbersome to use.)
